internal/alias/interceptor.go

```diff
@@ -0,0 +1,151 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: BUSL-1.1
+
+package alias
+
+import (
+	"context"
+	"strings"
+	"sync"
+
+	"github.com/hashicorp/boundary/internal/daemon/controller/handlers"
+	pbs "github.com/hashicorp/boundary/internal/gen/controller/api/services"
+	"github.com/hashicorp/boundary/sdk/pbs/controller/protooptions"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/reflect/protoreflect"
+	"google.golang.org/protobuf/reflect/protoregistry"
+	"google.golang.org/protobuf/types/descriptorpb"
+)
+
+func init() {
+	// register all proto fields that are aliasable.
+	protoregistry.GlobalTypes.RangeMessages(func(m protoreflect.MessageType) bool {
+		registerAliasableFields(m.Descriptor())
+		return true
+	})
+}
+
+var globalAliasableRegistry sync.Map
+
+func registerAliasableFields(d protoreflect.MessageDescriptor) {
+	fields := d.Fields()
+	for i := 0; i < fields.Len(); i++ {
+		f := fields.Get(i)
+		name := f.FullName()
+		opts, ok := f.Options().(*descriptorpb.FieldOptions)
+		if !ok {
+			continue
+		}
+		isAliasable := proto.GetExtension(opts, protooptions.E_Aliasable).(bool)
+		if isAliasable {
+			globalAliasableRegistry.Store(name, struct{}{})
+		}
+	}
+}
+
+// ResolutionInterceptor returns a grpc.UnaryServerInterceptor that resolves
+// alias values in the request to their corresponding destination ids. If no
+// alias is found or the alias has no destination id, an error is returned.
+// For an field in the request to be considered for alias resolution, it must
+// be annotated with the Aliasable proto option.
+func ResolutionInterceptor(
```

Comment:
I don't think any of this interceptor implementation should be in the domain package. This causes the domain layer to import things like `internal/daemon/controller/handlers` and `internal/gen/controller/api/services` which reverses the dependency order.

It might be ok to leave the bulk of the interceptor logic here as a domain service, if we can eliminate these imports. It looks like `services` is only imported to handle the special case for AuthorizeSession, see my other comment about that which might eliminate the need for that import. And it looks like `handlers` is only needed for constructing errors. That could be eliminated by a thin wrapper of the domain service to create a grpc interceptor in another layer, which applies formatting of the errors using the `handlers` package.